server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,256 @@
+
+    return cnt
+
+def verify_controllers(hierarchy, config):
+    cnt=0
+    h = os.path.basename(hierarchy)
+    archive = config.ARCHIVE
+    users = config.USERS
+
+    if h == 'incoming':
+        kind="incoming"
+        user=""
+    elif h == 'results':
+        kind="results"
+        user=""
+    elif os.path.dirname(hierarchy) == users:
+        kind="results"
+        user=hierarchy
+    else:
+        print("${PROG}: verify_controllers bad argument, hierarchy_root=\"${hierarchy_root}\"\n")
+        return 1
+
+    controllers = glob.iglob(os.path.join(hierarchy,'*'))
+    unexpected_objs, unexpected_objects, mialist, empty = ([] for i in range(4))
+
+    for controller in controllers:
+        c = os.path.basename(controller)
+        if os.path.isdir(controller):
+            if not os.path.isdir(os.path.join(archive,c)):
+                mialist.append(c)
+            else:
+                states = glob.iglob(os.path.join(controller,'*'))
+                for s in states: 
+                    if os.path.isdir(s): 
+                        if len(os.listdir(s))==0:
+                            empty.append(os.path.basename(s))
+                    elif not os.path.isdir(s) and not os.path.islink(s):
+                        unexpected_objects.append(c)
+        else:
+            unexpected_objs.append(c)
+
+    if len(unexpected_objs)>0:
+        cnt = output_format(unexpected_objs, "Unexpected files found:", cnt)
+    if len(mialist)>0:
+        cnt = output_format(mialist, "Controllers which do not have a ${ARCHIVE} directory:", cnt)
+    if len(empty)>0:
+        cnt = output_format(empty, "Controllers which are empty:", cnt)
+    if len(unexpected_objects)>0:
+        cnt = output_format(unexpected_objects, "Controllers which have unexpected objects:", cnt)
+
+    return cnt
+
+def verify_users(config):
+    cnt = 0
+    users = config.USERS
+
+    users_dirs = glob.iglob(os.path.join(users,'*'))
+    user_dir, unexpected_objects = ([] for i in range(2))
+
+    for user in users_dirs:
+        u = os.path.basename(user)
+        if os.path.isdir(user):
+            user_dir.append(u)
+            cnt = verify_controllers(user,config)
+            if cnt>0:
+                cnt = cnt+1
+        else:
+            unexpected_objects.append(u)
+
+    if len(unexpected_objects)>0:
+        cnt = output_format(unexpected_objects, "Unexpected files found:", cnt)
+    
+    return cnt
+
+def output_format(n, msg, cnt):
+    print("\t* ",msg)
+    print("\t  ++++++++++")
+    print(*n, sep='\n\t')
+    print("\t  ----------")
+    return cnt+1
+
+def verify_archive(config):
+
+    archive = config.ARCHIVE
+    cnt=0
+    controllers = glob.iglob(os.path.join(archive,'*'))
+    bad_controllers=[]
+
+    for controller in controllers:
+        if os.path.isdir(controller):
+            print("\nController: ", os.path.basename(controller))
+            cnt = verify_subdirs(config, controller, cnt)
```

Comment:
It is probably better *not* to pass count as an argument: the routine can count how many tings went wrong during its execution and return that as its value. The caller can keep a cumulattive count: when the called routine returns, it takes its return value and adds it to the cumulative count. That way, there is less linkage between the caller and the called routine.